ReSharper.FSharp/src/FSharp.Psi.Features/src/Intentions/FunctionAnnotationAction.fs

```diff
@@ -0,0 +1,136 @@
+        
+        // Tooltips in this case are prepended with "val " which we don't want in the source code - in the case that
+        // a type can't be determined, it doesn't start with val, so don't try to annotate type
+        let tooltipString = showL overload.MainDescription
+        
+        if tooltipString.IndexOf("val ", 0) = 0 then
+            (showL overload.MainDescription |> Seq.skip 4 |> Seq.map string) |> String.concat "" |> Some
+        else
+            None
+    override x.Text = "Annotate function with parameter types and return type"
+    
+    override x.IsAvailable _ =
+        let binding = dataProvider.GetSelectedElement<IBinding>()
+        if isNull binding then false else
+
+        match binding.HeadPattern.As<INamedPat>() with
+        | null -> false
+        | namedPat ->
+        isNotNull namedPat.Identifier
+        
+    override x.ExecutePsiTransaction(_, _) =
+        let binding = dataProvider.GetSelectedElement<IBinding>()
+        use _writeCookie = WriteLockCookie.Create(binding.IsPhysical())
+        use _disableFormatter = new DisableCodeFormatter()
+        match binding.HeadPattern.As<INamedPat>() with
+        | null -> null
+        | namedPat ->
+            
+        match box (namedPat.GetFSharpSymbol()) with
+        | null -> null
+        | methodSymbol ->
+        let fSharpFunction =
+            match methodSymbol with
+            | :? FSharpMemberOrFunctionOrValue as x -> x
+            | _ -> failwith "Expected function here"
+            
+        match namedPat.As<IParametersOwnerPat>() with
+        | null -> null
+        | parameterOwner ->
+        let treeParameters = parameterOwner.Parameters |> Seq.toList
+            
+        // Annotate function parameters
+        let factory = namedPat.CreateElementFactory()
+        let childrenToModify =
+            treeParameters
+            |> Seq.choose(
+                function
+                | :? ILocalReferencePat as ref ->
+                    Some {ParameterNameNode = ref; ParameterNodeInSignature = ref}
+                | :? IParenPat as ref ->
+                    let parameterNameNode =
+                        ref.Children()
+                        |> Seq.choose(function | :? ILocalReferencePat as pat -> Some pat | _ -> None)
+                        |> Seq.tryExactlyOne
+                    parameterNameNode
+                    |> Option.map(fun namedNode ->
+                        {ParameterNameNode = namedNode;
+                          ParameterNodeInSignature = ref})
+                | _ -> None)
+            |> Seq.toList
+            
+        // Parse FSharp file, in order to get tooltips
+        let parsedFile =
+            namedPat.FSharpFile.GetParseAndCheckResults(true, opName)
+            |> Option.defaultWith(fun () -> failwith "Unable to parse FSharp file")
+        let checkResults = parsedFile.CheckResults
+        let document = namedPat.FSharpFile.GetSourceFile().Document
```

Comment:
Reusing symbols from `GetFSharpSymbol` and `GetFSharpSymbolUse` should be enough here, no need to request ParseAndCheck results.